mmdet3d/models/dense_heads/primitive_head.py

```diff
@@ -0,0 +1,816 @@
+            point_boundary_offset_xy = torch.zeros([num_points, 3]).to(device)
+            point_boundary_sem_xy = torch.zeros(
+                [num_points, 3 + self.num_dim + 1]).to(device)
+        elif self.primitive_mode == 'line':
+            point_line_mask = torch.zeros(num_points).to(device)
+            point_line_offset = torch.zeros([num_points, 3]).to(device)
+            point_line_sem = torch.zeros([num_points, 3 + 1]).to(device)
+        else:
+            NotImplementedError
+
+        instance_flag = torch.nonzero(
+            pts_semantic_mask != self.num_classes).squeeze(1)
+        instance_labels = pts_instance_mask[instance_flag].unique()
+
+        for i, i_instance in enumerate(instance_labels):
+            ind = instance_flag[pts_instance_mask[instance_flag] == i_instance]
+            x = points[ind, :3]
+
+            # Corners
+            corners = gt_bboxes_3d.corners[i][[0, 1, 3, 2, 4, 5, 7, 6]]
+            xmin, ymin, zmin = corners.min(0)[0]
+            xmax, ymax, zmax = corners.max(0)[0]
+
+            # Get lower four lines
+            plane_lower_temp = torch.as_tensor([0, 0, 1,
+                                                -corners[6, -1]]).to(device)
+            para_points = corners[[1, 3, 5, 7]]
+            newd = torch.sum(para_points * plane_lower_temp[:3], 1)
+            if self.check_upright(para_points) and \
+                    plane_lower_temp[0] + plane_lower_temp[1] < \
+                    self.train_cfg['lower_thresh']:
+                plane_lower = torch.as_tensor([0, 0, 1, plane_lower_temp[-1]
+                                               ]).to(device)
+                plane_upper = torch.as_tensor([0, 0, 1,
+                                               -torch.mean(newd)]).to(device)
+            else:
+                import pdb
+                pdb.set_trace()
+                print('error with upright')
+            if self.check_z(plane_upper, para_points) is False:
+                import pdb
+                pdb.set_trace()
+
+            # Get the boundary points here
+            alldist = torch.abs(
+                torch.sum(x * plane_lower[:3], 1) + plane_lower[-1])
+            mind = alldist.min()
+            sel = torch.abs(alldist - mind) < self.train_cfg['dist_thresh']
+
+            # Get lower four lines
+            line_sel1, line_sel2, line_sel3, line_sel4 = self.get_linesel(
+                x[sel], xmin, xmax, ymin, ymax)
+            if self.primitive_mode == 'line':
+                if torch.sum(line_sel1) > self.train_cfg['num_point_line']:
+                    point_line_mask[ind[sel][line_sel1]] = 1.0
+                    linecenter = torch.mean(x[sel][line_sel1], axis=0)
+                    linecenter[1] = (ymin + ymax) / 2.0
+                    point_line_offset[
+                        ind[sel][line_sel1]] = linecenter - x[sel][line_sel1]
+                    point_line_sem[ind[sel][line_sel1]] = torch.as_tensor([
```

Comment:
For all the usage of ` torch.as_tensor(data).to(device)`, we can use `points.new_tensor(data)` for simipilicity